src/wgetopt.rs

```diff
@@ -25,341 +26,337 @@ Cambridge, MA 02139, USA.  */
             optstring = &optstring[1..];
         }
 
         self.shortopts = optstring;
         self.initialized = true;
     }
 
-    /// Advance to the next ARGV-element.
-    /// \return Some(\0) on success, or None or another value if we should stop.
-    fn _advance_to_next_argv(&mut self) -> Option<char> {
-        let argc = self.argc();
-        if self.ordering == Ordering::PERMUTE {
-            // If we have just processed some options following some non-options, exchange them so
-            // that the options come first.
-            if self.first_nonopt != self.last_nonopt && self.last_nonopt != self.woptind {
-                self.exchange();
-            } else if self.last_nonopt != self.woptind {
-                self.first_nonopt = self.woptind;
+    /// Advance to the next element in `argv`. Returns nothing on success, or
+    /// an optional value if we need to stop.
+    fn next_argv(&mut self) -> Result<(), Option<char>> {
+        let argc = self.argv.len();
+
+        if self.ordering == Ordering::Permute {
+            // Permute the args if we've found options following non-options.
+            if self.last_nonopt != self.wopt_index {
+                if self.first_nonopt == self.last_nonopt {
+                    self.first_nonopt = self.wopt_index;
+                } else {
+                    self.exchange();
+                }
             }
 
-            // Skip any additional non-options and extend the range of non-options previously
-            // skipped.
-            while self.woptind < argc
-                && (self.argv[self.woptind].char_at(0) != '-' || self.argv[self.woptind].len() == 1)
+            // Skip any further non-options, adjusting the relevant indices as
+            // needed.
+            while self.wopt_index < argc
+                && (self.argv[self.wopt_index].char_at(0) != '-'
+                    || self.argv[self.wopt_index].len() == 1)
             {
-                self.woptind += 1;
+                self.wopt_index += 1;
             }
-            self.last_nonopt = self.woptind;
+
+            self.last_nonopt = self.wopt_index;
         }
 
-        // The special ARGV-element `--' means premature end of options. Skip it like a null option,
-        // then exchange with previous non-options as if it were an option, then skip everything
-        // else like a non-option.
-        if self.woptind != argc && self.argv[self.woptind] == "--" {
-            self.woptind += 1;
+        // The `--` element prevents any further scanning of options. We skip it like
+        // a null option, then swap with non-options as if were an option, then skip
+        // everything else like a non-option.
+        if self.wopt_index != argc && self.argv[self.wopt_index] == "--" {
+            self.wopt_index += 1;
 
-            if self.first_nonopt != self.last_nonopt && self.last_nonopt != self.woptind {
+            if self.first_nonopt == self.last_nonopt {
```

Comment:
The original code read:

```rust
           if self.first_nonopt != self.last_nonopt && self.last_nonopt != self.woptind {
                self.exchange();
            } else if self.first_nonopt == self.last_nonopt {
                self.first_nonopt = self.woptind;
            }
```

so this has dropped a condition. I can be convinced that `self.last_nonopt != self.woptind` will always be true, given that we just incremented `woptind`; but in that case we should assume it's true in both branches.

Reply:
In the original code, the condition `self.last_nonopt != self.woptind` is only checked if `self.first_nonopt != self.last_nonopt` is true -- which is by definition equivalent to `self.first_nonopt == self.last_nonopt` being false.

Therefore, checking `self.first_nonopt == self.last_nonopt` and only checking `self.last_nonopt != self.woptind` if it's false should be logically equivalent to the original.